File: taurus-protect-sdk-python/taurus_protect/mappers/token_metadata.py
"""Token metadata mapper for converting OpenAPI DTOs to domain models."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Optional

from taurus_protect.mappers._base import safe_string
from taurus_protect.models.token_metadata import (
    CryptoPunkMetadata,
    FATokenMetadata,
    TokenMetadata,
)

if TYPE_CHECKING:
    pass  # For OpenAPI types when available


def token_metadata_from_dto(dto: Any) -> Optional[TokenMetadata]:
    """
    Convert OpenAPI TgvalidatordERCTokenMetadata to domain TokenMetadata.

    Args:
        dto: OpenAPI token metadata DTO.

    Returns:
        Domain TokenMetadata model or None if dto is None.
    """
    if dto is None:
        return None

    return TokenMetadata(
        name=safe_string(getattr(dto, "name", None)),
        description=safe_string(getattr(dto, "description", None)),
        decimals=safe_string(getattr(dto, "decimals", None)),
        uri=safe_string(getattr(dto, "uri", None)),
        data_type=safe_string(getattr(dto, "data_type", None)),
        base64_data=safe_string(getattr(dto, "base64_data", None)),
    )


def fa_token_metadata_from_dto(dto: Any) -> Optional[FATokenMetadata]:
    """
    Convert OpenAPI TgvalidatordFATokenMetadata to domain FATokenMetadata.

    Args:
        dto: OpenAPI FA token metadata DTO.

    Returns:
        Domain FATokenMetadata model or None if dto is None.
    """
    if dto is None:
        return None

    return FATokenMetadata(
        name=safe_string(getattr(dto, "name", None)),
        symbol=safe_string(getattr(dto, "symbol", None)),
        decimals=safe_string(getattr(dto, "decimals", None)),
        description=safe_string(getattr(dto, "description", None)),
        uri=safe_string(getattr(dto, "uri", None)),
        data_type=safe_string(getattr(dto, "data_type", None)),
        base64_data=safe_string(getattr(dto, "base64_data", None)),
    )


def crypto_punk_metadata_from_dto(dto: Any) -> Optional[CryptoPunkMetadata]:
    """
    Convert OpenAPI TgvalidatordCryptoPunkMetadata to domain CryptoPunkMetadata.

    Args:
        dto: OpenAPI CryptoPunk metadata DTO.

    Returns:
        Domain CryptoPunkMetadata model or None if dto is None.
    """
    if dto is None:
        return None

    return CryptoPunkMetadata(
        punk_index=safe_string(getattr(dto, "punk_index", None)),
        image_url=safe_string(getattr(dto, "image_url", None)),
        attributes=getattr(dto, "attributes", None),
    )

File: taurus-protect-sdk-go/pkg/protect/mapper/token_metadata.go
package mapper

import (
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/internal/openapi"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// CryptoPunkMetadataFromDTO converts an OpenAPI CryptoPunkMetadata to a domain CryptoPunkMetadata.
func CryptoPunkMetadataFromDTO(dto *openapi.TgvalidatordCryptoPunkMetadata) *model.CryptoPunkMetadata {
	if dto == nil {
		return nil
	}

	return &model.CryptoPunkMetadata{
		PunkID:     safeString(dto.PunkId),
		Attributes: safeString(dto.PunkAttributes),
		Image:      safeString(dto.Image),
	}
}

// ERCTokenMetadataFromDTO converts an OpenAPI ERCTokenMetadata to a domain ERCTokenMetadata.
func ERCTokenMetadataFromDTO(dto *openapi.TgvalidatordERCTokenMetadata) *model.ERCTokenMetadata {
	if dto == nil {
		return nil
	}

	return &model.ERCTokenMetadata{
		Name:        safeString(dto.Name),
		Description: safeString(dto.Description),
		Decimals:    safeString(dto.Decimals),
		DataType:    safeString(dto.DataType),
		Base64Data:  safeString(dto.Base64Data),
		URI:         safeString(dto.Uri),
	}
}
